Cleanup: Use rustc's `same_type` for our `same_tys`

phansch opened the pull request:
This delegates our `same_tys` to [ty::TyS::same_type][same_type] to
remove some duplication.



Our `same_tys` was introduced 4 years ago in #730. Before, it was
building an inference context to be able to call
`can_eq` to compare the types. The `rustc-dev-guide` has some more details
about `can_eq` in [Type Inference -> Trying equality][try_eq].

Now, using the rustc function, we are essentially comparing the `DefId`s
of the given types, which also makes more sense, IMO.

I also confirmed that the FIXME is resolved via a bit of `dbg!`, however
no UI tests seem to have been affected.

[same_type]: https://github.com/rust-lang/rust/blob/659951c4a0d7450e43f61c61c0e87d0ceae17087/src/librustc_middle/ty/util.rs#L777
[try_eq]: https://rustc-dev-guide.rust-lang.org/type-inference.html#trying-equality

---

changelog: none

phansch commented:
Ah, it might make sense to also check whether or not the FIXME was resolved already without this PR. Will do that tomorrow.

flip1995 commented on clippy_lints/src/utils/mod.rs:
```diff
@@ -890,17 +890,8 @@ pub fn return_ty<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, fn_item: hir::HirId) -> T
-        .infer_ctxt()
-        .enter(|infcx| infcx.can_eq(cx.param_env, a, b).is_ok())
+pub fn same_tys<'tcx>(a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
+    ty::TyS::same_type(a, b)
```
I think this function should be completely removed and `TyS;;sane_type` (imported) used everywhere.

phansch commented:
Confirmed that the FIXME was already resolved without these changes:
![Selection_078](https://user-images.githubusercontent.com/2042399/80316911-eaddbf80-8800-11ea-9a99-b18a6606f518.png)

I removed our `same_tys` and updated all call sites to use the rustc function instead

flip1995 commented:
Hmm, generally `TyS` should never be used outside of the `ty` module. This should be enforced by an internal rustc lint. But this lint is one of the lints, I wanted to fix for quite some time. Also `same_type` would have to be re-exported to prevent the usage of `TyS` in Clippy.

phansch commented:
> Hmm, generally TyS should never be used outside of the ty module.

A right, I'm going to give re-exporting a try later today :+1:

bors commented:
:umbrella: The latest upstream changes (presumably #5583) made this pull request unmergeable. Please resolve the merge conflicts.

flip1995 commented:
@phansch Did you have any success with reexporting?

phansch commented:
> @phansch Did you have any success with reexporting?

Not really. `same_type` is the only associated _function_ of `TyS`. Everything else is a method and accessible via instances of `Ty`. As far as I can tell the function would either have to be turned into a method or moved out of `TyS`?

flip1995 commented:
> As far as I can tell the function would either have to be turned into a method or moved out of TyS?

We can just keep it as it is right now, since the internal rustc lint doesn't trigger here. So I would be fine with this right now. (This again reminds me, that I wanted to revisit this internal rustc lint...)

flip1995 commented:
so rebase + CI pass = r=me

phansch commented:
@bors r=flip1995

bors commented:
:pushpin: Commit 623faac84ec56fa545163ab81d7e3b759a392353 has been approved by `flip1995`

<!-- @bors r=flip1995 623faac84ec56fa545163ab81d7e3b759a392353 -->
<!-- homu: {"type":"Approved","sha":"623faac84ec56fa545163ab81d7e3b759a392353","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 623faac84ec56fa545163ab81d7e3b759a392353 with merge e6f05ad9125beee9e43bf75d9d3742efc72e8f99...
<!-- homu: {"type":"BuildStarted","head_sha":"623faac84ec56fa545163ab81d7e3b759a392353","merge_sha":"e6f05ad9125beee9e43bf75d9d3742efc72e8f99"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/744921856), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/744920891), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/744956061)
Approved by: flip1995
Pushing e6f05ad9125beee9e43bf75d9d3742efc72e8f99 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/744956061","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/744921856","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/744920891"},"merge_sha":"e6f05ad9125beee9e43bf75d9d3742efc72e8f99"} -->
